jlm/llvm/opt/alias-analyses/MemoryStateEncoder.cpp

```diff
@@ -390,6 +408,35 @@ class RegionalizedStateMap final {
   }
 
 private:
+  jlm::rvsdg::output&
+  GetOrInsertUndefinedMemoryState(jlm::rvsdg::region & region)
+  {
+    return HasUndefinedMemoryState(region)
+           ? GetUndefinedMemoryState(region)
+           : InsertUndefinedMemoryState(region);
+  }
+
+  bool
+  HasUndefinedMemoryState(const jlm::rvsdg::region & region) const noexcept
+  {
+    return UndefinedMemoryStates_.find(&region) != UndefinedMemoryStates_.end();
+  }
+
+  jlm::rvsdg::output&
+  GetUndefinedMemoryState(const jlm::rvsdg::region & region) const noexcept
+  {
+    JLM_ASSERT(HasUndefinedMemoryState(region));
+    return *UndefinedMemoryStates_.find(&region)->second;
+  }
+
+  jlm::rvsdg::output&
+  InsertUndefinedMemoryState(jlm::rvsdg::region & region) noexcept
```

Comment:
Wouldn't it make sense to check for if an undefinedMemoryState already exists?

Reply:
That is what the `GetOrInsertUndefinedMemoryState` method 20 lines above does, no? It calls this method. Or do I misunderstand your comment?

Reply:
Yes, GetOrInsertUndefinedMemoryState checks if one exists before calling insert, so if this is the only way that insert is called, then the undefined state is only created once.
But, if one instead call insert multiple times, then one would get a new undefined state for each time.
So maybe a JLM_ASSERT could be useful.

Reply:
`GetUndefinedMemoryState()`, `InsertUndefinedMemoryState()`, and `GetOrInsertUndefinedMemoryState()` are all `private` in the `RegionalizedStateMap` class. `GetUndefinedMemoryState()` and `InsertUndefinedMemoryState()` are only called in `GetOrInsertUndefinedMemoryState()`. Thus, all calls go through `GetOrInsertUndefinedMemoryState()`, which means it is ensured that only one undefined state node per region is created.